lib/simple-options.nix

```
@@ -0,0 +1,89 @@
+{
+  lib,
+  attrAttr ? "attrsOf",
+  listAttr ? "listOf",
+  subMAttr ? "options",
+  typeAttr ? "type",
+  debug    ? false,
+  ...
+}:
+module:
+let
+  mkOptionW = breadcrumb: optName: optDef:
+    let result = lib.mkOption (toOption breadcrumb optName optDef);
+    in
+      if debug then
+        builtins.trace ''
+          function: modulazy.mkOptionWrapper
+          args:
+            breadcrumb: ${builtins.concatStringsSep "." breadcrumb}
+            optName: ${optName}
+            optDef: ${builtins.concatStringsSep " " (builtins.attrNames optDef)}
+          result: ${builtins.concatStringsSep " " (builtins.attrNames result)}
+        ''
+        result
+      else result;
+
+  toOptions = breadcrumb: options: { options = builtins.mapAttrs (mkOptionW breadcrumb) options; };
+
+  toType    = breadcrumb: type:
+    let result =
+      if builtins.typeOf breadcrumb != "list"
+        then builtins.throw ''breadcrumb must be a list, instead it is a ${builtins.typeOf breadcrumb}'' else
+      if builtins.typeOf type       != "set"
+        then builtins.throw ''type must be a attrset, instead it is a ${builtins.typeOf type}, breadcrumb: ${builtins.concatStringsSep "." breadcrumb}'' else
+      if builtins.hasAttr "_type" type
+        then type else
```

```suggestion
      type._type or (
```

This is also a micro-optimization: fewer function calls.
I have to admit that I haven't measured it though.